### PR TITLE
Feature/platoon strategic plugin

### PR DESCRIPTION
**Pull request by saina-ramyar**
# PR Details
## Description

Added platooning strategic plugin

## Related Issue
#738 

## Motivation and Context

The platooning strategic plugin is required to plan maneuver and communicate with other platoon vehicles.

## How Has This Been Tested?
Unit tested. Needs to be integration tested.

## Types of changes

- [ ] Defect fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

- [x] I have added any new packages to the sonar-scanner.properties file
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
[CARMA Contributing Guide](https://github.com/usdot-fhwa-stol/carma-platform/blob/develop/Contributing.md) 
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: platooning_strategic/src/main.cpp [hunk 1]
```diff
@@ -0,0 +1,11 @@
+
```
Comment (kjrush) [inline platooning_strategic/src/main.cpp:1]: CARMA copyright notice

File: platooning_strategic/src/platoon_strategic.cpp [hunk 2]
```diff
@@ -0,0 +1,420 @@
...
+        ros::CARMANodeHandle::spin();
+
+    }
+
+    void PlatoonStrategicPlugin::pose_cb(const geometry_msgs::PoseStampedConstPtr& msg)
+    {
+        pose_msg_ = msg;
+    }
+
+    // TODO: fill this function
```
Comment (kjrush) [inline platooning_strategic/src/platoon_strategic.cpp:113]: Does this still need to be done? If so, it'd be good to leave a bit more detailed comment about what needs to be implemented.

File: platooning_strategic/src/platoon_strategic.cpp [hunk 3]
```diff
@@ -0,0 +1,420 @@
...
+    bool PlatoonStrategicPlugin::plan_trajectory_cb(cav_srvs::PlanTrajectoryRequest &req, cav_srvs::PlanTrajectoryResponse &resp){
+        
+
+        return true;
+    }
+
+
+    void PlatoonStrategicPlugin::run_standby() const {
+
```
Comment (kjrush) [inline platooning_strategic/src/platoon_strategic.cpp:122]: Empty, if this is used in a spin-loop be sure that we're not busy-waiting on this or we'll spin as fast as possible which'll waste CPU cycles.

File: platooning_strategic/test/test_platoon_manager.cpp [hunk 4]
```diff
@@ -0,0 +1,121 @@
+#include "platoon_manager.hpp"
```
Comment (kjrush) [inline platooning_strategic/test/test_platoon_manager.cpp:1]: CARMA copyright notice

File: platooning_strategic/src/platoon_manager.cpp [hunk 5]
```diff
@@ -0,0 +1,406 @@
+#include "platoon_manager.hpp"
+#include <boost/algorithm/string.hpp>
+#include <ros/ros.h>
+#include <array>
+
+
+namespace platoon_strategic
+{
+    PlatoonManager::PlatoonManager(){}
```
[the remark this answers is not in the document]
Comment (saina-ramyar) [reply platooning_strategic/src/platoon_manager.cpp:24]: created that instructor to use in unit tests without node handle.

File: platooning_strategic/src/platoon_manager.cpp [hunk 6]
```diff
@@ -0,0 +1,406 @@
...
+
+    PlatoonManager::PlatoonManager(ros::NodeHandle *nh): nh_(nh) {
```
[the remark this answers is not in the document]
Comment (saina-ramyar) [reply platooning_strategic/src/platoon_manager.cpp]: changed to shared_ptr

File: platooning_strategic/include/state_machine.hpp [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+        double maxAllowedJoinGap = 90;
+        int maxPlatoonSize = 10;
+        double vehicleLength = 5.0;
+        std::mutex plan_mutex_;
+        int infoMessageInterval;
+        std::string targetPlatoonId;
+        std::string OPERATION_INFO_TYPE = "INFO";
+        std::string OPERATION_STATUS_TYPE = "STATUS";
+        std::string JOIN_AT_REAR_PARAMS = "SIZE:%1%,SPEED:%2%,DTD:%3%";
+        std::string  MOBILITY_STRATEGY = "Carma/Platooning";
```
[the remark this answers is not in the document]
Comment (saina-ramyar) [reply platooning_strategic/include/state_machine.hpp]: done

File: platooning_strategic/include/platoon_manager.hpp [hunk 8]
```diff
@@ -0,0 +1,162 @@
+#pragma once
```
[the remark this answers is not in the document]
Comment (saina-ramyar) [reply platooning_strategic/include/platoon_manager.hpp:1]: added

File: platooning_strategic/src/platoon_manager.cpp [hunk 9]
```diff
@@ -0,0 +1,406 @@
...
+    PlatoonMember PlatoonManager::getLeader(){
+        PlatoonMember* leader = new PlatoonMember("", "", 0.0, 0.0, 0.0, 0.0);
+        if(isFollower && platoon.size() != 0) {
+            // return the first vehicle in the platoon as default if no valid algorithm applied
+            *leader = platoon[0];
+            if (algorithmType == "APF_ALGORITHM"){
+                while (!ros::isShuttingDown()){
+                    int newLeaderIndex = allPredecessorFollowing();
+                    if(newLeaderIndex < platoon.size() && newLeaderIndex >= 0) {
+
```
[the remark this answers is not in the document]
Comment (saina-ramyar) [reply platooning_strategic/src/platoon_manager.cpp]: yes, it was missed. Added the logic

File: platooning_strategic/src/platoon_manager.cpp [hunk 10]
```diff
@@ -0,0 +1,406 @@
...
+    double PlatoonManager::getPlatoonRearDowntrackDistance(){
+        if(platoon.size() == 0) {
+            double dist = getCurrentDowntrackDistance();
+            return dist;
+        }
+        return platoon[platoon.size() - 1].vehiclePosition;
+    }
+
+    PlatoonMember PlatoonManager::getLeader(){
+        PlatoonMember* leader = new PlatoonMember("", "", 0.0, 0.0, 0.0, 0.0);
```
[the remark this answers is not in the document]
Comment (saina-ramyar) [reply platooning_strategic/src/platoon_manager.cpp]: could not find a place to free, so removed the pointer

File: platooning_strategic/src/platoon_manager.cpp [hunk 11]
```diff
@@ -0,0 +1,406 @@
...
+                ROS_DEBUG("    Location = "                                , pm.vehiclePosition);
+                ROS_DEBUG("    CommandSpeed = "                            , pm.commandSpeed);
+                isExisted = true;
+                break;
+            }
+        }
+
+        if(!isExisted) {
+            long cur_t = ros::Time::now().toSec()*1000;
+            PlatoonMember* newMember = new PlatoonMember(senderId, senderBsmId, cmdSpeed, curSpeed, dtDistance, cur_t);
```
[the remark this answers is not in the document]
Comment (saina-ramyar) [reply platooning_strategic/src/platoon_manager.cpp]: removed the pointer

File: platooning_strategic/src/platoon_manager.cpp [hunk 12]
```diff
@@ -0,0 +1,406 @@
...
+        return platoon[platoon.size() - 1].vehiclePosition;
+    }
+
+    PlatoonMember PlatoonManager::getLeader(){
+        PlatoonMember* leader = new PlatoonMember("", "", 0.0, 0.0, 0.0, 0.0);
+        if(isFollower && platoon.size() != 0) {
+            // return the first vehicle in the platoon as default if no valid algorithm applied
+            *leader = platoon[0];
+            if (algorithmType == "APF_ALGORITHM"){
+                while (!ros::isShuttingDown()){
```
[the remark this answers is not in the document]
Comment (saina-ramyar) [reply platooning_strategic/src/platoon_manager.cpp]: removed that.

File: platooning_strategic/src/platoon_manager.cpp [hunk 13]
```diff
@@ -0,0 +1,406 @@
...
+            speed[i] = platoon[i].vehicleSpeed;
+        }
+        double cur_speed = getCurrentSpeed();
+        speed[speed.size() - 1] = cur_speed;
+
+        ///***** Case Two *****///
+        // If the distance headway between the subject vehicle and its predecessor is an issue
+        // according to the "min_gap" and "max_gap" thresholds, then it should follow its predecessor
+        // The following line will not throw exception because the length of downtrack array is larger than two in this case
+        double timeHeadwayWithPredecessor = downtrackDistance[downtrackDistance.size() - 2] - downtrackDistance[downtrackDistance.size() - 1];
```
[the remark this answers is not in the document]
Comment (saina-ramyar) [reply platooning_strategic/src/platoon_manager.cpp:203]: I used the variable names from CARMA2. It is called time headway but it actually means gap (distance).

File: platooning_strategic/src/platoon_manager.cpp [hunk 14]
```diff
@@ -0,0 +1,406 @@
...
+        // Generate an array of speed for every vehicles in this platoon including the host vehicle
+        // The size of speed array is platoon.size() + 1, because the platoon list did not contain the host vehicle
+        std::vector<double> speed(platoon.size() + 1);
+        for(int i = 0; i < platoon.size(); i++) {
+            speed[i] = platoon[i].vehicleSpeed;
+        }
+        double cur_speed = getCurrentSpeed();
+        speed[speed.size() - 1] = cur_speed;
+
+        ///***** Case Two *****///
```
[the remark this answers is not in the document]
Comment (saina-ramyar) [reply platooning_strategic/src/platoon_manager.cpp:199]: Again, copied from carma2. I think case 1 is an edge case.

File: platooning_strategic/src/platoon_manager.cpp [hunk 15]
```diff
@@ -0,0 +1,406 @@
...
+    std::vector<double> PlatoonManager::getTimeHeadwayFromIndex(std::vector<double> timeHeadways, int start) const {
+        std::vector<double> result(timeHeadways.begin() + start-1, timeHeadways.end());
+        return result;
+    }
+    
+
+    bool PlatoonManager::insufficientGapWithPredecessor(double distanceToFrontVehicle) {
+        bool frontGapIsTooSmall = distanceToFrontVehicle < minGap;
+        bool previousLeaderIsPredecessor = (previousFunctionalLeaderID == platoon[platoon.size() - 1].staticId);
+        bool frontGapIsNotLargeEnough = (distanceToFrontVehicle < maxGap && previousLeaderIsPredecessor);
```
[the remark this answers is not in the document]
Comment (saina-ramyar) [reply platooning_strategic/src/platoon_manager.cpp:299]: Got the name from carma2. Dont understand what is weird about it?:)

File: platooning_strategic/src/platoon_manager.cpp [hunk 16]
```diff
@@ -0,0 +1,406 @@
...
+        return (frontGapIsTooSmall || frontGapIsNotLargeEnough);
+    }
+
+    std::vector<double> PlatoonManager::calculateTimeHeadway(std::vector<double> downtrackDistance, std::vector<double> speed) const{
+        std::vector<double> timeHeadways(downtrackDistance.size() - 1);
+        for (int i=0; i<timeHeadways.size(); i++){
+            if (speed[i+1]!=0){
+                timeHeadways[i] = (downtrackDistance[i] - downtrackDistance[i + 1]) / speed[i + 1];
```
[the remark this answers is not in the document]
Comment (saina-ramyar) [reply platooning_strategic/src/platoon_manager.cpp:307]: The list is not sorted, but those are positions of vehicles in platoon, so the difference does not become negative

File: platooning_strategic/src/platoon_manager.cpp [hunk 17]
```diff
@@ -0,0 +1,406 @@
...
+        if(isFollower) {
+            return platoon.size();
+        }
+        return 0;
+    }
+
+    
+
+    double PlatoonManager::getDistanceFromRouteStart() const{
+        return 0.0;
```
[the remark this answers is not in the document]
Comment (saina-ramyar) [reply platooning_strategic/src/platoon_manager.cpp:379]: Yes, it is not used now. But will keep it for more testing

File: platooning_strategic/include/state_machine.hpp [hunk 18]
```diff
@@ -0,0 +1,114 @@
+#pragma once
```
[the remark this answers is not in the document]
Comment (saina-ramyar) [reply platooning_strategic/include/state_machine.hpp:1]: added

File: platooning_strategic/include/state_machine.hpp [hunk 19]
```diff
@@ -0,0 +1,114 @@
...
+
+        // PlatooningStateMachine();
+
+        // PlatooningStateMachine(ros::NodeHandle *nh);
+
+
+        MobilityRequestResponse onMobilityRequestMessage(cav_msgs::MobilityRequest &msg);
+        void onMobilityResponseMessage(cav_msgs::MobilityResponse &msg);
+        void onMobilityOperationMessage(cav_msgs::MobilityOperation &msg);
+        cav_msgs::Maneuver planManeuver(double current_dist, double end_dist, double current_speed, double target_speed, int lane_id, ros::Time current_time);
```
[the remark this answers is not in the document]
Comment (saina-ramyar) [reply platooning_strategic/include/state_machine.hpp]: added

File: platooning_strategic/include/state_machine.hpp [hunk 20]
```diff
@@ -0,0 +1,114 @@
...
+        
+        PlatoonState current_platoon_state;
+        std::string applicantID;
+        PlatoonPlan current_plan;
+
+        std::string targetLeaderId = "";
+
+        PlatoonManager *pm_;
```
[the remark this answers is not in the document]
Comment (saina-ramyar) [reply platooning_strategic/include/state_machine.hpp]: removed the pointer

File: platooning_strategic/src/state_machine.cpp [hunk 21]
```diff
@@ -0,0 +1,453 @@
+#include "state_machine.hpp"
```
[the remark this answers is not in the document]
Comment (saina-ramyar) [reply platooning_strategic/src/state_machine.cpp:16]: added

File: platooning_strategic/src/state_machine.cpp [hunk 22]
```diff
@@ -0,0 +1,453 @@
...
+            break;
+        
+        default:
+            throw std::invalid_argument("State machine attempting to process an illegal state value");
+        }
+        
+    }
+
+    cav_msgs::Maneuver PlatooningStateMachine::planManeuver(double current_dist, double end_dist, double current_speed, double target_speed, int lane_id, ros::Time current_time){
+
```
[the remark this answers is not in the document]
Comment (saina-ramyar) [reply platooning_strategic/src/state_machine.cpp:136]: filled it

File: platooning_strategic/src/state_machine.cpp [hunk 23]
```diff
@@ -0,0 +1,453 @@
+#include "state_machine.hpp"
+
+namespace platoon_strategic
+{
+
+    // PlatooningStateMachine::PlatooningStateMachine(){} ???????????
```
[the remark this answers is not in the document]
Comment (saina-ramyar) [reply platooning_strategic/src/state_machine.cpp]: fixed

File: platooning_strategic/src/state_machine.cpp [hunk 24]
```diff
@@ -0,0 +1,453 @@
...
+    cav_msgs::Maneuver PlatooningStateMachine::planManeuver(double current_dist, double end_dist, double current_speed, double target_speed, int lane_id, ros::Time current_time){
+        
+    }
+
+
+
+
+    MobilityRequestResponse PlatooningStateMachine::onMobilityRequestMessageFollower(cav_msgs::MobilityRequest &msg) const
+    {
+        ROS_DEBUG("Initialized all node handles");
```
[the remark this answers is not in the document]
Comment (saina-ramyar) [reply platooning_strategic/src/state_machine.cpp]: That shouldnt have been there. revomved

File: platooning_strategic/src/state_machine.cpp [hunk 25]
```diff
@@ -0,0 +1,453 @@
...
+        else{
+            ROS_DEBUG("Received platoon request with vehicle id = " , msg.header.sender_id);
+            ROS_DEBUG("The request type is " , msg.plan_type.type,  " and we choose to ignore");
+            return MobilityRequestResponse::NO_RESPONSE;
+        }
+    }
+
+    void PlatooningStateMachine::onMobilityResponseMessageLeaderWaiting(cav_msgs::MobilityResponse &msg) const
+    {
+
```
[the remark this answers is not in the document]
Comment (saina-ramyar) [reply platooning_strategic/src/state_machine.cpp]: commented.

File: platooning_strategic/src/state_machine.cpp [hunk 26]
```diff
@@ -0,0 +1,453 @@
...
+
+    MobilityRequestResponse PlatooningStateMachine::onMobilityRequestMessageStandby(cav_msgs::MobilityRequest &msg) const
+    {
+        // In standby state, the plugin is not responsible for replying to any request messages
+        return MobilityRequestResponse::NO_RESPONSE;
+    }
+    void PlatooningStateMachine::onMobilityResponseMessageStandby(cav_msgs::MobilityResponse &msg) const
+    {
+
```
[the remark this answers is not in the document]
Comment (saina-ramyar) [reply platooning_strategic/src/state_machine.cpp]: commented

File: platooning_strategic/src/platoon_strategic.cpp (hunk 2 above)
Comment (saina-ramyar) [reply platooning_strategic/src/platoon_strategic.cpp]: That function is removed

File: platooning_strategic/src/main.cpp (hunk 1 above)
Comment (saina-ramyar) [reply platooning_strategic/src/main.cpp:15]: added

File: platooning_strategic/include/platoon_strategic.hpp [hunk 27]
```diff
@@ -0,0 +1,170 @@
...
+            int LEADER_TIMEOUT_COUNTER_LIMIT = 5;
+            double waitingStateTimeout = 25.0; // s
+            double desiredJoinGap = 30.0; // m
+            double desiredJoinTimeGap = 4.0; // s
+
+
+            std::string MOBILITY_STRATEGY = "Carma/Platooning";
+            std::string OPERATION_INFO_TYPE = "INFO";
+            std::string OPERATION_STATUS_TYPE = "STATUS";
+            std::string OPERATION_STATUS_PARAMS = "STATUS|CMDSPEED:%1%,DTD:%2%,SPEED:%3%";
```
[the remark this answers is not in the document]
Comment (saina-ramyar) [reply platooning_strategic/include/platoon_strategic.hpp]: made const

File: platooning_strategic/include/platoon_strategic.hpp [hunk 28]
```diff
@@ -0,0 +1,170 @@
...
+            void run_standby() const;
+            void run_leader();
+            void run_leader_waiting();
+            void run_candidate_follower();
+            void run_follower();
+
+        
+        private:
+
+            PlatooningStateMachine *psm_;
```
[the remark this answers is not in the document]
Comment (saina-ramyar) [reply platooning_strategic/include/platoon_strategic.hpp]: modified

File: platooning_strategic/src/state_machine.cpp [hunk 29]
```diff
@@ -0,0 +1,453 @@
...
+                if (current_plan.valid){
+                    bool isForCurrentPlan = (msg.header.plan_id == current_plan.planId);
+                    bool isFromTargetVehicle = (msg.header.sender_id == targetLeaderId);
+                    if(isForCurrentPlan && isFromTargetVehicle) {
+                        if(msg.is_accepted) {
+                        // We change to follower state and start to actually follow that leader
+                        // The platoon manager also need to change the platoon Id to the one that the target leader is using 
+                        ROS_DEBUG("The leader " , msg.header.sender_id , " agreed on our join. Change to follower state.");
+                        pm_->changeFromLeaderToFollower(targetPlatoonId);
+                        current_platoon_state = PlatoonState::FOLLOWER;
```
[the remark this answers is not in the document]
Comment (saina-ramyar) [reply platooning_strategic/src/state_machine.cpp:429]: Right! I dont think that will happen. I am not sure how to modify this, but let me know if you have a suggestion

File: platooning_strategic/test/test_platoon_manager.cpp (hunk 4 above)
Comment (saina-ramyar) [reply platooning_strategic/test/test_platoon_manager.cpp:1]: added

File: platooning_strategic/src/platoon_manager.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (kjrush) [reply platooning_strategic/src/platoon_manager.cpp:203]: Okay, if that's consistent then it's fine for now.

File: platooning_strategic/src/platoon_manager.cpp (hunk 17 above)
[the remark this answers is not in the document]
Comment (kjrush) [reply platooning_strategic/src/platoon_manager.cpp:379]: Cool, that's fine. I'd just recommend adding a comment to indicate as such, just to be clear that it's not the intended implementation.

File: platooning_strategic/src/state_machine.cpp [hunk 30]
```diff
@@ -106,8 +112,29 @@ namespace platoon_strategic
...
+            maneuver_msg.lane_following_maneuver.parameters.planning_tactical_plugin = "PlatooningTacticalPlugin";
+            maneuver_msg.lane_following_maneuver.parameters.planning_strategic_plugin = "PlatooningStrategicPlugin";
+            maneuver_msg.lane_following_maneuver.start_dist = start_dist;
+            maneuver_msg.lane_following_maneuver.start_speed = pm_.command_speed_;
+            maneuver_msg.lane_following_maneuver.start_time = current_time;
+            maneuver_msg.lane_following_maneuver.end_dist = start_dist + (mvr_duration_*pm_.command_speed_);
+            maneuver_msg.lane_following_maneuver.end_speed = pm_.command_speed_;
+            // because it is a rough plan, assume vehicle can always reach to the target speed in a lanelet
+            maneuver_msg.lane_following_maneuver.end_time = current_time + ros::Duration(mvr_duration_);
+            maneuver_msg.lane_following_maneuver.lane_id = "";
```
Comment (kjrush) [inline platooning_strategic/src/state_machine.cpp:134]: I'd recommend adding a "//TODO: Populate actual lane ID" comment here, so we don't lose track of this.

File: platooning_strategic/src/state_machine.cpp (hunk 29 above)
[the remark this answers is not in the document]
Comment (kjrush) [reply platooning_strategic/src/state_machine.cpp:429]: On second pass, it shouldn't matter unless we're multi-threading this node. By default ROS is single-threaded, so it shouldn't be a problem.

**Pull request merged**